types/mempool/sender_nonce.go

```diff
@@ -0,0 +1,199 @@
+package mempool
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/signing"
+	huandu "github.com/huandu/skiplist"
+)
+
+var (
+	_ Mempool  = (*senderNonceMempool)(nil)
+	_ Iterator = (*nonceMempoolIterator)(nil)
+)
+
+type senderTxs struct {
+	txQueue *huandu.SkipList
+	head    *huandu.Element
+}
+
+func newSenderTxs() senderTxs {
+	return senderTxs{
+		head:    nil,
+		txQueue: huandu.New(huandu.LessThanFunc(txKeyLessNonce)),
+	}
+}
+
+func (s *senderTxs) insert(key txKey, tx sdk.Tx) {
+	s.txQueue.Set(key, tx)
+	s.head = s.txQueue.Front()
+}
+
+func (s *senderTxs) getMove() *huandu.Element {
+	if s.head == nil {
+		return nil
+	}
+	currentHead := s.head
+	s.head = s.head.Next()
+	return currentHead
+}
+
+func (s *senderTxs) remove(key txKey) error {
+	res := s.txQueue.Remove(key)
+	if res == nil {
+		return ErrTxNotFound
+	}
+	if s.head == res {
+		s.head = s.txQueue.Front()
+	}
+	return nil
+}
+
+type senderNonceMempool struct {
+	senders map[string]*senderTxs
+	rnd     *rand.Rand
+}
+
+func NewSenderNonceMempool() Mempool {
+	senderMap := make(map[string]*senderTxs)
+	snp := &senderNonceMempool{
+		senders: senderMap,
+	}
+	snp.setSeed(time.Now().UnixNano())
```

Comment:
Let have the default be seed be from crypto/rand instead of system time?

```golang
	var seed int64
	binary.Read(crand.Reader, binary.BigEndian, &seed)
	return mrand.NewSource(seed)
```

Reply:
True I forgot about it